server/src/main/java/org/elasticsearch/gateway/GatewayService.java

```diff
@@ -215,10 +215,8 @@ private void performStateRecovery(final boolean enforceRecoverAfterTime, final S
                 threadPool.generic().execute(new AbstractRunnable() {
                     @Override
                     public void onFailure(final Exception e) {
-                        logger.warn("Recovery failed", e);
-                        // we reset `recovered` in the listener don't reset it here otherwise there might be a race
-                        // that resets it to false while a new recover is already running?
-                        GatewayService.this.onFailure("state recovery failed: " + e.getMessage());
+                        logger.warn("state recovery failed", e);
+                        resetRecoveredFlags();
```

Comment:
Should we `resetRecoveredFlags()` on failure of the scheduled task a few lines above this too?

Reply:
sure, added in 9de5c0e